crates/api/src/op_store.rs

```diff
@@ -0,0 +1,86 @@
+//! Kitsune2 op store types.
+
+use crate::{K2Result, OpId, Timestamp};
+use futures::future::BoxFuture;
+use std::cmp::Ordering;
+use std::sync::Arc;
+
+/// An op with metadata.
+///
+/// This is the basic unit of data in the kitsune2 system.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct MetaOp {
+    /// The id of the op.
+    pub op_id: OpId,
+
+    /// The creation timestamp of the op.
+    ///
+    /// This must be the same for everyone who sees this op.
+    ///
+    /// The host must reject the op if the timestamp does not agree with any timestamps inside the
+    /// op data.
+    pub timestamp: Timestamp,
+
+    /// The actual op data.
+    pub op_data: Vec<u8>,
+
+    /// Crdt-style add-only opaque implementor-use flags.
+    pub op_flags: std::collections::HashSet<String>,
+}
+
+impl Ord for MetaOp {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (&self.timestamp, &self.op_id).cmp(&(&other.timestamp, &other.op_id))
+    }
+}
+
+impl PartialOrd for MetaOp {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// The API that a kitsune2 host must implement to provide data persistence for kitsune2.
+pub trait OpStore: 'static + Send + Sync + std::fmt::Debug {
+    /// Process incoming ops.
+    ///
+    /// Pass the incoming ops to the host for processing. The host is expected to store the ops
+    /// if it is able to process them.
```

Comment:
This would be coming from the network, so we can't trust the ops being ordered by timestamp is correct. Plus it's not so important because the ops have to be merged into the host's list which will involve inserting them into an ordered index of some kind whether they are sorted or not. It would be slightly more efficient if they were sorted.. but not very. Oh and also, we can't have the timestamps here, that needs to be part of the op data and not the metaop, so at this point we wouldn't even know how to sort them in Kitsune.